src/NUnitEngine/nunit.engine/Services/Tokenizer.cs

```diff
@@ -0,0 +1,241 @@
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+// 
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+using System.Text;
+using NUnit.Common;
+
+namespace NUnit.Engine
+{
+    public enum TokenKind
+    {
+        Eof,
+        Word,
+        String,
+        Symbol
+    }
+
+    public class Token
+    {
+        public Token(TokenKind kind) : this(kind, string.Empty) { }
+
+        public Token(TokenKind kind, char ch) : this(kind, ch.ToString()) { }
+
+        public Token(TokenKind kind, string text)
+        {
+            Kind = kind;
+            Text = text;
+        }
+
+        public TokenKind Kind { get; private set; }
+
+        public string Text { get; private set; }
+
+        #region Equality Overrides
+
+        public override bool Equals(object obj)
+        {
+            return obj is Token && this == (Token)obj;
+        }
+
+        public override int GetHashCode()
+        {
+            return Text.GetHashCode();
+        }
+
+        public override string ToString()
+        {
+            return Text != null
+                ? Kind.ToString() + ":" + Text
+                : Kind.ToString();
+        }
+
+        public static bool operator ==(Token t1, Token t2)
+        {
+            bool t1Null = ReferenceEquals(t1, null);
+            bool t2Null = ReferenceEquals(t2, null);
+
+            if (t1Null && t2Null)
+                return true;
+
+            if (t1Null || t2Null)
+                return false;
+
+            return t1.Kind == t2.Kind && t1.Text == t2.Text;
+        }
+
+        public static bool operator !=(Token t1, Token t2)
+        {
+            return !(t1 == t2);
+        }
+
+        #endregion
+    }
+
+    /// <summary>
+    /// Tokenizer class performs lexical analysis for the TestSelectionParser.
+    /// It recognizes a very limited set of tokens: words, symbols and
+    /// quoted strings. This is sufficient for the simple DSL we use to
+    /// select which tests to run.
+    /// </summary>
+    public class Tokenizer
+    {
+        private string _input;
+        private int _index;
+
+        private const char EOF = '\0';
+        private const string WORD_CHARS = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_-.";
```

Comment:
I'm using them for both now. The rhs can generally be either a word or a string.